service/history/historyEngine.go

```diff
@@ -294,6 +302,56 @@ func (e *historyEngineImpl) GetWorkflowExecutionNextEventID(
 		RunId:      request.Execution.RunId,
 	}
 
+	response, err := e.getWorkflowExecutionNextEventID(domainID, execution)
+	if err != nil {
+		return nil, err
+	}
+
+	// expectedNextEventID is 0 when caller want to get the current next event ID without blocking
+	expectedNextEventID := request.GetExpectedNextEventId()
+
+	// if caller decide to long poll on workflow execution
+	// and the event ID we are looking for is smaller than current next event ID
+	if expectedNextEventID >= response.GetEventId() && response.GetIsWorkflowRunning() {
+		subscriberID, channel, err := e.historyEventNotifier.WatchHistoryEvent(&domainID, &execution)
+		if err != nil {
+			return nil, err
+		}
+		defer e.historyEventNotifier.UnwatchHistoryEvent(&domainID, &execution, subscriberID)
+
+		// check again in case the next event ID is updated
+		response, err = e.getWorkflowExecutionNextEventID(domainID, execution)
+		if err != nil {
+			return nil, err
+		}
+
+		if expectedNextEventID < response.GetEventId() || !response.GetIsWorkflowRunning() {
+			return response, nil
+		}
+
+		timeoutChan := time.NewTimer(e.shard.GetConfig().LongPollExpirationInterval).C
```

Comment:
Can you instead use the deadline on context?

Reply:
i actually used two channel for indication of "done"
one is this timer, waiting 20s (less than request timeout, so we can always return meaningful result), another one is "case <-ctx.Done():", ref:https://github.com/uber/cadence/blob/46b0e459b892d401f188f4eaacfaa640c5b6659a/service/history/historyEngine.go#L343